### PR TITLE
[Net] asmap to improve IP bucketing in addrman - backports

### PR DESCRIPTION
**Pull request by furszy**
Decoupled from #2411, built on top of #2479. Probably the last decouple from the "road to Tor" work.

Focused on porting the ASN nodes bucketing functionality. The hearth of this work is #16702.

Providing an asmap file that contains the IP->ASN mapping, nodes will be bucketed by AS they belong to, in order to make impossible for a node to connect to several nodes hosted in a single AS.
This is done in response to Erebus attack, but also to generally diversify the connections every node creates, especially useful when a large fraction of nodes operate under a couple of cloud providers.

#### List of PRs:
* #7932
* #10765
* #13532
* #13575
* #16702
* #17812
* #18023
* #19314

PRs for a follow up PR:
* #18029
* #18512

### CODE REVIEW

**Comment by furszy** [comment]
rebased on master, ready to go.

File: test/functional/feature_asmap.py [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+            self.stop_node(0)
+            with self.node.assert_debug_log(expected_messages(self.default_asmap)):
+                self.start_node(0, [arg])
+        os.remove(self.default_asmap)
+
+    def test_default_asmap_with_missing_file(self):
+        self.log.info('Test bitcoind -asmap with missing default map file')
+        self.stop_node(0)
+        msg = "Error: Could not find or parse specified asmap: '\"{}\"'".format(self.default_asmap)
+        self.node.assert_start_raises_init_error(extra_args=['-asmap'], expected_msg=msg)
```
Comment (random-zebra) [inline test/functional/feature_asmap.py:79]: to fix this test case:
```suggestion
        self.assert_start_raises_init_error(0, extra_args=['-asmap'], expected_msg=msg)
```

File: test/functional/feature_asmap.py [hunk 2]
```diff
@@ -66,18 +67,27 @@ def test_asmap_with_relative_path(self):
...
         msg = "Error: Could not find asmap file '\"{}\"'".format(self.default_asmap)
         self.node.assert_start_raises_init_error(extra_args=['-asmap'], expected_msg=msg)
 
+    def test_empty_asmap(self):
+        self.log.info('Test pivxd -asmap with empty map file')
+        self.stop_node(0)
+        with open(self.default_asmap, "w", encoding="utf-8") as f:
+            f.write("")
+        msg = "Error: Could not parse asmap file \"{}\"".format(self.default_asmap)
+        self.node.assert_start_raises_init_error(extra_args=['-asmap'], expected_msg=msg)
```
Comment (random-zebra) [inline test/functional/feature_asmap.py:87]: to fix this test case as well:
```suggestion
        self.assert_start_raises_init_error(0, extra_args=['-asmap'], expected_msg=msg)
```

**Review by random-zebra — COMMENTED** [review]
Light code review ACK (with the last 2 test cases of `feature_asmap` fixed).
Will do some more testing.

File: test/functional/feature_asmap.py (hunk 1 above)
Comment (furszy) [reply test/functional/feature_asmap.py]: nice! ☕

**Comment by furszy** [comment]
Updated per zebra's feedback 👌, fixed the two commented test cases :).

**Review by random-zebra — APPROVED** [review]
Tested (also with `demo.map` from sipa's repo) ACK 97e04978971d163c44e518739818af50c5bc6562

Nit:
Commit 63b6f995991c3d9a3aa3cd522ea73d0b6c634757 should be split/squashed into 1e9a8568e4948fdc6a172efcd9bc9cd844cdff94 + a69b82d4e7427419087f5847bc7ac999e8a64633

**Comment by furszy** [comment]
done, squashed. I missed it between all of the open works.

**Review by random-zebra — DISMISSED** [review]
utACK a7138cfa150e8c3743bbe98588208a1eab436063

**Comment by Fuzzbawls** [comment]
IMHO, I think eb3eba03ebca443f254cea39465620b811ff8eb8 should be dropped and the issue fixed directly in upstream.

The extra indentation goes against the styling policies/guidelines for both us and upstream, and I would prefer to avoid intentionally (re-)introducing a policy violation that had been explicitly fixed in the past for a marginal improvement in ease of cherry-picking, especially being relatively close to upstream's HEAD in this particular file; which, given a quick glance, seems to be one of, if not the only file that has this styling discrepancy.

**Comment by Fuzzbawls** [comment]
Upstream PR addressing the indentation issue can be found at <https://github.com/bitcoin/bitcoin/pull/22562>

**Comment by furszy** [comment]
yeah, it's the only file with text indentation discrepancies.

I made it because after, literally, more than two hundred commits back ported and adapted to our sources, it was becoming a pain in the ass to struggle against the same text indentation issue over and over, (with literally zero benefits as it's just text indentation that can be re-added all at once at the end). Conflicting most of the file most of the times.

Saying that, now that the road to Tor path is clear (In `netaddress.h` history this is at 12 commits from the HEAD in #2411), it shouldn't take me much to drop the indentation change and apply the changes to every posterior commit. Will give it a quick look. It's one of those "meh" changes that no one in upstream really cared about for all of this years that we can have first.

**Comment by furszy** [comment]
done, #2411 rebase took me a bit more but all looking good there too. Nothing to worry about.

**Comment by furszy** [comment]
rebased again, conflict solved. Ready to go.

**Review by random-zebra — APPROVED** [review]
re-utACK 16791f2a412af58d1f5186fc8095a8c362398286

**Review by Fuzzbawls — APPROVED** [review]
ACK 16791f2a412af58d1f5186fc8095a8c362398286

**Pull request merged**